### PR TITLE
qmpi: Add initial support for QMPI

### PR DESCRIPTION
**Pull request by wesbland**
This work is based on the QMPI work in the MPI Forum Tools WG (see https://github.com/mpiwg-tools/tools-issues/issues/1) and is another implementation of the work done in the QMPI paper published at EuroMPI 2019 (see https://doi.org/10.1145/3343211.3343215) and in the reference implementation provided (see https://github.com/caps-tum/qmpi).

The work in this branch is similar to that initial implementation, but is intended to explore implementing QMPI inside the MPI implementation as opposed to through the external PMPI interface. Optimizations will be added as the work continues.

#### Initialization / Finalization

The tool should use a system specific option to initialize the library when it's loaded. GCC and other compiler use the `__attribute__((constructor)` trick to accomplish this. The same can be used during finalization.

#### Registration

During the initialization call, the tool should call:

```c
int QMPI_Register_tool(void * tool_context, int * tool_id);
```

This function allows the tool to provide a context object (`tool_context`) which lets the tool store some information specific to its incarnation of itself (in case there are multiple copies of a tool). MPI will return the `tool_id` to indicate where in the stack this tool has been registered. Tools will be called in reverse order (i.e., the largest number will be called first and will move backward until MPICH is called). MPICH will always be tool 0.

Then the tool can register each function that it wishes to override with this call:

```c
int QMPI_Register_function(int calling_tool_id, enum QMPI_Functions_enum function_enum, void (* function_ptr)(void))
```

This function takes `calling_tool_id` as the ID of the tool that is registering functions and `function_enum` as a way of specifying which MPI function is being registered. The enum values are defined in `mpi.h and here:

```c
enum QMPI_Functions_enum {
    MPI_ABORT_T,
    MPI_ACCUMULATE_T,
    ... snip ...
    MPI_WTIME_T,
    MPI_LAST_FUNC
};
```

`function_ptr` should be a pointer to the tool's interception function pointer. This function pointer will be the same signature as the MPI function, with an addition `void *` argument in the last position to allow tools to get their context objects back at call time. These functions are all `typedef`ed in `mpi.h` in the form of `QMPI_<function_name>_t` (e.g. `QMPI_Send_t`).

#### Interception

When the tool's interception function is called, it can get any tool-specific information from the context object. It is recommended that the tool stores its ID in that object for easy retrieval later. When the tool is done, it should call this function to determine the next tool in the QMPI stack (which the tool is responsible for calling):

```c
int QMPI_Get_function(int calling_tool_id, enum QMPI_Functions_enum function_enum, void (** function_ptr)(void), void **next_tool_context);
```

This function also uses `calling_tool_id` and `function_enum` to determine the current tool's ID and the enum value for the function being queried. It also provides a function pointer in a similar format to the registration function. Finally, the function returns the context object of the next tool in the call stack which should be used when calling the returned function.

An example of what a tool that implements this API would look like is here: https://gist.github.com/wesbland/065de74b3582913c3945099af5cf5130

## Expected Impact

Some basic benchmarks with OSU MBW and Latency show about 3% or less performance loss with one tool attached.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

File: src/mpi/init/get_function_table.c [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+addition to the MPI library itself.
+
+   Notes:
+
+.N Fortran
+
+.N Errors
+.N MPI_SUCCESS
+@*/
+QMPI_Function_pointers_t **QMPIX_Get_function_table(void)
```
Comment (wesbland) [inline src/mpi/init/get_function_table.c:24]: @pavanbalaji made the point that this function might not be necessary. We could provide the pointer to the function table when calling the registration function at init time.

File: src/mpi/pt2pt/recv.c [hunk 2]
```diff
@@ -61,6 +61,17 @@ length of the message can be determined with 'MPI_Get_count'.
 @*/
 int MPI_Recv(void *buf, int count, MPI_Datatype datatype, int source, int tag,
              MPI_Comm comm, MPI_Status * status)
+{
+    /* Call the first layer in the QMPI stack that contains the correct function. */
+    for (int i = 0; i <= MPIR_QMPI_num_tools; i++) {
```
Comment (wesbland) [inline src/mpi/pt2pt/recv.c:66]: There's an obvious optimization to be added here when `MPIR_QMPI_num_tools == 0` to avoid the function pointers.

**Comment by wesbland** [comment]
Added a link in the description to a Gist implementing this API: https://gist.github.com/wesbland/065de74b3582913c3945099af5cf5130

**Comment by pavanbalaji** [comment]
What's the right namespace for this?  The `QMPI_Register_callbacks` (which can probably just be renamed to `XXX_Init`) is not defined by the MPI library, but by the tool.  So `MPI_` doesn't seem like the right namespace.

**Comment by wesbland** [comment]
> What's the right namespace for this?  The `QMPI_Register_callbacks` (which can probably just be renamed to `XXX_Init`) is not defined by the MPI library, but by the tool.  So `MPI_` doesn't seem like the right namespace.

The reason I named it this way is because this function will need to be defined in the standard so MPI knows what to look for and tools know what to provide. I don't care what the final name is.

**Comment by wesbland** [comment]
# Original Implementation Notes

To use this functionality, the tool needs to provide a function called QMPI_Register_callbacks which returns a struct of function pointers that the tool wishes to override (currently only send and recv are supported). The tool must be compiled as a dynamically linked library and at runtime, when running the application, the environment variable `QMPI_TOOL_LIST` should include the name of the .so file containing the tool. This can either be the name of the tool with the library in the `LD_LIBRARY_PATH`, or a full path to the library file.

At runtime, MPICH will search for the tool with dlopen, call the indicated function, and store the function pointers provided. When the user calls an MPI function, MPICH will find the first entry in the function pointer list and call that function (which will either be from the tool or the MPI library itself). The tool is responsible for calling the function `QMPI_Get_function_table` to get the full table of function pointers and calling the next function in the table via the provided input value of layer_id. Eventually, this will lead to the MPICH-level function being called.

At finalization time, the tool is responsible for cleaning up its own layer of the function table stack by freeing the memory and then calling the next layer of the QMPI stack.

**Comment by wesbland** [comment]
# 2020-08-27 Update

Based on feedback from the Tools WG, the following changes were made:

* Correct the use of function pointers to use void(*)(void) instead of `void *`.
* Add a tool-defined context object that is provided to MPI on registration.
* Use a system-specific way of registering the tool rather than using `dlopen` and `dlsym` (e.g. using `_attribute__((constructor)))`.
* Create an MPI-level function to register each tool function rather that using a struct of function pointers.
* Create an MPI-level function to get the next function in the stack rather than querying the array of structs of function pointers.
* Create an MPI-exposed enum to let tools specify which MPI function they are referencing when registering or querying a function.
* Create MPI-exposed typedefs for each function to allow the tool to use them if desired.

**Comment by wesbland** [comment]
# 2020-09-18 Update

## User-facing changes

* Added support for all MPI functions with some caveats:
    * I haven't tried anything with Fortran. I explicitly excluded those functions when generating the bindings except where they didn't exist as C bindings.
    * My current implementation breaks the build with ROMIO disabled. I'll work to fix that later.
* The tools are now required to register each individual function that they want to intercept.
* When the tool is intercepting a function, it is required to query for the next function in the list (rather than getting a table of function pointers during setup time).

## Internal Changes

* The scripts to generate the bindings are both included for now (`maint/qmpi_binding.py` and `qmpi.py`). The former should probably hang around. The later maybe should not? Or maybe they need to be merged together?
* The QMPI bindings are in `qmpi.h` which is included in `mpi.h`. `qmpi.h` is also included in the install directory for MPICH.
    * This meant that a few build system changes were necessary, particularly in ROMIO to look in the `main_top_srcdr` when looking for this header.
* The new functions to support QMPI described in the top of this PR are no longer in the `QMPIX` namespace. They are now `QMPI` functions. This is also true for all of the `QMPI` functions inside MPICH.
    * It might not make sense for these to be the namespaces long term, but they're not breaking any rules so we can figure this out later.
* The enum parameters now use `enum QMPI_Functions_enum` instead of `int` types since that better reflects what they are.

**Comment by wesbland** [comment]
This PR has now been replaced by https://github.com/pmodels/mpich/pull/5035

**Pull request closed without merging**